include/onnxruntime/core/common/gpu_profiler_common.h

```diff
@@ -0,0 +1,450 @@
+    auto const& client_handle_offset = client_it->second;
+    auto external_correlation = unique_correlation_id - client_handle_offset.second;
+    auto& event_list = per_client_events_by_ext_correlation_[client_handle_offset.first][external_correlation];
+    return &event_list;
+  }
+
+  void MapEventsToClient(uint64_t unique_correlation_id, std::vector<EventRecord>&& events) {
+    auto p_event_list = GetEventListForUniqueCorrelationId(unique_correlation_id);
+    if (p_event_list != nullptr) {
+      p_event_list->insert(p_event_list->end(),
+                           std::make_move_iterator(events.begin()),
+                           std::make_move_iterator(events.end()));
+    }
+  }
+
+  void DeferEventMapping(EventRecord&& event, uint64_t tracer_correlation_id) {
+    events_pending_client_mapping_[tracer_correlation_id].emplace_back(std::move(event));
+  }
+
+  std::mutex manager_instance_mutex_;
+  uint64_t next_client_id_ = 1;
+  uint64_t num_active_clients_ = 0;
+  bool tracing_enabled_ = false;
+  std::mutex unprocessed_activity_buffers_mutex_;
+  std::mutex activity_buffer_processor_mutex_;
+
+  // Unprocessed activity buffers
+  std::vector<ProfilerActivityBuffer> unprocessed_activity_buffers_;
+
+  // Keyed on unique_correlation_id -> (client_id/client_handle, offset)
+  // unique_correlation_id - offset == external_correlation_id
+  InlinedHashMap<uint64_t, std::pair<uint64_t, uint64_t>> unique_correlation_id_to_client_offset_;
+
+  // Keyed on tracer_correlation_id -> unique_correlation_id
+  InlinedHashMap<uint64_t, uint64_t> tracer_correlation_to_unique_correlation_;
+
+  // client_id/client_handle -> external_correlation_id -> events
+  InlinedHashMap<uint64_t, std::map<uint64_t, Events>> per_client_events_by_ext_correlation_;
+
+  // Keyed on tracer correlation_id, keeps track of activity records
+  // for which we haven't established the external_correlation_id yet.
+  InlinedHashMap<uint64_t, std::vector<EventRecord>> events_pending_client_mapping_;
+}; /* class GPUTracerManager */
+
+// Base class for a GPU profiler
+template <typename TManager>
+class GPUProfilerBase : public EpProfiler {
+ protected:
+  GPUProfilerBase() = default;
+  virtual ~GPUProfilerBase() {}
+
+  void MergeEvents(std::map<uint64_t, Events>& events_to_merge, Events& events) {
+    Events merged_events;
+
+    auto event_iter = std::make_move_iterator(events.begin());
+    auto event_end = std::make_move_iterator(events.end());
+    for (auto& map_iter : events_to_merge) {
+      auto ts = static_cast<long long>(map_iter.first);
+
+      // find the last occurence of a matching timestamp,
+      // if one exists
+      while (event_iter != event_end &&
+             (event_iter->ts < ts ||
+              (event_iter->ts == ts &&
+               (event_iter + 1) != event_end &&
+               (event_iter + 1)->ts == ts))) {
+        merged_events.emplace_back(*event_iter);
+        ++event_iter;
+      }
```

Comment:
It seems that you are doing a linear scan here, does `events` guaranteed to be sorted by timestamp?

Reply:
From my understanding of the code, `events` is sorted by timestamp. That said, I've only refactored existing code, which assumed that `events` is sorted in non-decreasing order.